File: src/zammad/zammad.py
import logging

import requests

from zammad.call import CallZammad


class Zammad:
    endpoint: str

    def __init__(self, endpoint: str):
        self.endpoint = endpoint

    def new_call(self, call: CallZammad):
        call_from, _, direction = self.parse_call(call)
        payload = {
            "event": "newCall",
            "from": call_from,
            "to": "",
            "user[]": call.agent_name,
            "direction": direction,
            "call_id": call.callid,
            "callid": call.callid,
        }

        return self.make_request(payload)

    def hangup(self, call: CallZammad):
        call_from, call_to, direction = self.parse_call(call)
        payload = {
            "event": "hangup",
            "from": call_from,
            "to": call_to,
            "direction": direction,
            "call_id": call.callid,
            "callid": call.callid,
            "cause": call.cause,
            "answeringNumber": call.agent
        }

        return self.make_request(payload)

    def answer(self, call: CallZammad):
        call_from, call_to, direction = self.parse_call(call)
        payload = {
            "event": "answer",
            "from": call_from,
            "to": call_to,
            "direction": direction,
            "call_id": call.callid,
            "callid": call.callid,
            "answeringNumber": call.agent
        }
        if call.direction == 'Inbound':
            payload['user[]'] = call.agent_name

        return self.make_request(payload)

    def make_request(self, payload: dict):
        resp = requests.post(self.endpoint, json=payload)
        try:
            if resp.status_code == 200:
                logging.info("Event sent to Zammad")
            else:
                logging.error("Error sending Event to Zammad: " + resp.text)
        except requests.RequestException as err:
            logging.error("Error sending Event to Zammad:" + str(err))

    def parse_call(self, call: CallZammad) -> [str, str, str]:
        if call.direction == "Inbound":
            call_from = call.number
            direction = "in"
            try:
                call_to = call.agent
            except:
                call_to = ""
                call.agent = ""
        else:
            call_from = call.agent
            call_to = call.number
            direction = "out"

        if not call.agent_name:
            call.agent_name = ''

        return [call_from, call_to, direction]

File: zammad.go
package zammadbridge

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"

	"github.com/rs/zerolog/log"
)

type ZammadApiRequest struct {
	Event           string `json:"event"`
	From            string `json:"from"`
	To              string `json:"to"`
	Direction       string `json:"direction"`
	CallId          string `json:"call_id"`
	CallIdDuplicate string `json:"callid"`
	Cause           string `json:"cause,omitempty"`
	AnsweringNumber string `json:"answeringNumber,omitempty"`
	User            string `json:"user,omitempty"`
}

// ZammadNewCall notifies Zammad that a new call came in. This is the
// first call required to process calls using Zammad.
func (z *ZammadBridge) ZammadNewCall(call *CallInformation) error {
	err := z.ZammadPost(ZammadApiRequest{
		Event:           "newCall",
		From:            call.CallFrom,
		To:              call.CallTo,
		Direction:       call.Direction,
		CallId:          call.CallUID,
		AnsweringNumber: call.AgentNumber,
		User:            call.AgentName,
	})
	call.ZammadInitialized = true
	if err != nil {
		return err
	}

	return nil
}

// ZammadAnswer notifies Zammad that the existing call was now answered by
// an agent.
func (z *ZammadBridge) ZammadAnswer(call *CallInformation) error {
	var user string
	if call.Direction == "Inbound" {
		user = call.AgentName
	}

	if !call.ZammadInitialized {
		err := z.ZammadNewCall(call)
		if err != nil {
			return fmt.Errorf("unable to initialize call with Zammad: %w", err)
		}
	}

	if call.ZammadAnswered {
		return nil // Nothing to do - TODO: can we redirect the call in Zammad?
	}

	err := z.ZammadPost(ZammadApiRequest{
		Event:           "answer",
		From:            call.CallFrom,
		To:              call.CallTo,
		Direction:       call.Direction,
		CallId:          call.CallUID,
		AnsweringNumber: call.AgentNumber,
		User:            user,
	})
	call.ZammadAnswered = true

	if err != nil {
		return err
	}

	return nil
}

// ZammadHangup notifies Zammad that the call was finished with a given cause.
// Possible values for `cause` are: "cancel", "normalClearing"
func (z *ZammadBridge) ZammadHangup(call *CallInformation, cause string) error {
	if !call.ZammadInitialized {
		err := z.ZammadNewCall(call)
		if err != nil {
			return fmt.Errorf("unable to initialize call with Zammad: %w", err)
		}
	}

	return z.ZammadPost(ZammadApiRequest{
		Event:           "hangup",
		From:            call.CallFrom,
		To:              call.CallTo,
		Direction:       call.Direction,
		CallId:          call.CallUID,
		Cause:           cause,
		AnsweringNumber: call.AgentNumber,
	})
}

// ZammadPost makes a POST Request to Zammad with the given payload
func (z *ZammadBridge) ZammadPost(payload ZammadApiRequest) error {
	// Processing
	if payload.Direction == "Inbound" {
		payload.Direction = "in"
	}
	if payload.Direction == "Outbound" {
		payload.Direction = "out"
	}
	payload.CallIdDuplicate = payload.CallId

	// Actual request
	requestBody, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("unable to serialize JSON request body: %w", err)
	}

	log.Trace().Str("call_id", payload.CallId).Str("event", payload.Event).Str("from", payload.From).Str("to", payload.To).Msg("Zammad request (POST)")
	resp, err := z.ClientZammad.Post(z.Config.Zammad.Endpoint, "application/json", bytes.NewBuffer(requestBody))
	if err != nil {
		return fmt.Errorf("unable to make request: %w", err)
	}

	log.Trace().Str("call_id", payload.CallId).Str("event", payload.Event).Str("from", payload.From).Str("to", payload.To).Int("status", resp.StatusCode).Msg("Zammad response (POST)")

	if resp.StatusCode >= 300 {
		data, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("unexpected response from Zammad (HTTP %d): %s", resp.StatusCode, string(data))
	}

	return nil
}
